File: svc-event/src/main/java/com/beancounter/event/service/EventSchedule.java
package com.beancounter.event.service;

import com.beancounter.auth.client.LoginService;
import com.beancounter.common.event.CorporateEvent;
import com.beancounter.common.utils.DateUtils;
import java.time.LocalDate;
import java.util.Collection;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@Slf4j
public class EventSchedule {

  private LoginService loginService;
  private final EventService eventService;
  private final DateUtils dateUtils = new DateUtils();

  public EventSchedule(EventService eventService) {
    this.eventService = eventService;
  }

  @Autowired(required = false)
  public void setLoginService(LoginService loginService) {
    this.loginService = loginService;
  }

  @Scheduled(cron = "${event.schedule:0 */30 7-18 ? * Tue-Sat}")
  void processEventsForRange() {
    log.info("Checking for corporate events to process");
    if (loginService != null) {
      loginService.login();
    }

    LocalDate end = dateUtils.getDate();
    LocalDate start = end.minusDays(5);
    Collection<CorporateEvent> events = eventService.findInRange(start, end);
    for (CorporateEvent event : events) {
      eventService.processMessage(event);
    }
    if (!events.isEmpty()) {
      log.info("Processed {} events", events.size());
    }
  }

}

File: svc-event/src/main/kotlin/com/beancounter/event/service/EventSchedule.kt
package com.beancounter.event.service

import com.beancounter.auth.client.LoginService
import com.beancounter.common.utils.DateUtils
import io.sentry.spring.jakarta.tracing.SentryTransaction
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service

/**
 * Scheduled event execution to obtain corporate events.
 */
@Service
class EventSchedule(
    private val eventService: EventService,
    private val dateUtils: DateUtils
) {
    private var loginService: LoginService? = null
    private val log = LoggerFactory.getLogger(EventSchedule::class.java)

    @Autowired(required = false)
    fun setLoginService(loginService: LoginService?) {
        this.loginService = loginService
    }

    @SentryTransaction(operation = "scheduled", name = "EventSchedule.processEventsForRange")
    @Scheduled(
        cron = "#{@eventsSchedule}",
        zone = "#{@scheduleZone}"
    )
    fun processEventsForRange() {
        val currentLoginService = loginService
        if (currentLoginService == null) {
            log.warn("LoginService not available, skipping event processing")
            return
        }

        // Simple approach: wrap entire operation in retry logic
        currentLoginService.retryOnJwtExpiry {
            // Authenticate
            currentLoginService.setAuthContext(currentLoginService.loginM2m())

            // Process events
            val end = dateUtils.date
            val start = end.minusDays(5)

            log.debug("Processing events from $start to $end")

            val events = eventService.findInRange(start, end)

            events.forEach { event ->
                eventService.processEvent(event)
            }

            if (events.isEmpty()) {
                log.info("No corporate events to process")
            } else {
                log.info("Successfully processed {} corporate events", events.size)
            }
        }
    }
}